content/advent-2015/test2doc-generate-your-api-docs.md

```diff
@@ -0,0 +1,294 @@
+```
+<br>
+
+`julienschmidt/httprouter`:
+
+```go
+import (
+	"net/http"
+
+	"github.com/adams-sarah/test2doc/doc/parse"
+	"github.com/julienschmidt/httprouter"
+)
+
+// MakeURLVarExtractor returns a func which extracts 
+// url vars from a request for test2doc documentation generation
+func MakeURLVarExtractor(router *httprouter.Router) parse.URLVarExtractor {
+	return func(req *http.Request) map[string]string {
+		// httprouter Lookup func needs a trailing slash on path
+		path := req.URL.Path
+		if !strings.HasSuffix(path, "/") {
+			path += "/"
+		}
+
+		_, params, ok := router.Lookup(req.Method, path)
+		if !ok {
+			return nil
+		}
+
+		paramsMap := make(map[string]string, len(params))
+		for _, p := range params {
+			paramsMap[p.Key] = p.Value
+		}
+
+		return paramsMap
+	}
+}
+
+// and then..
+test.RegisterURLVarExtractor(MakeURLVarExtractor(router))
+```
+<br>
+
+## How the magic happens
+
+### Recording Requests and Responses:
+Recording the requests and responses was pretty straight-forward:
+
+1. Write the HTTP handler's response initially to an `http.ResponseRecorder`
+2. Add the response header/body to the documentation
+3. Copy the response back to the original `http.ResponseWriter`
+4. Continue executing the test
+
+
+### Fetching the handler's Go doc string:
+First, let's take a look at the `http.ResponseWriter` interface
+from Go's `net/http` package:
+
+```go
+type ResponseWriter interface {
+        Header() Header
+        Write([]byte) (int, error)
+        WriteHeader(int)
+}
+```
+
+
+test2doc has its own `ResponseWriter`, which implements
+`http.ResponseWriter`, and looks something like this:
+
+```go
+type ResponseWriter struct {
+	HandlerInfo HandlerInfo
+	URLVars     map[string]string
+	W           *httptest.ResponseRecorder
+}
+```
+
+The test2doc `ResponseWriter`'s implementations of the `Header` and 
+`WriteHeader` methods just use those of its embedded 
+`httptest.ResponseRecorder`:
+
+```go
+func (rw *ResponseWriter) Header() http.Header {
+	return rw.W.Header()
+}
+
+func (rw *ResponseWriter) WriteHeader(c int) {
+	rw.W.WriteHeader(c)
+}
+```
+
+The magic is in the `Write` implementation. 
+Since every HTTP handler will call the `Write` method at 
+some point to write out the response, we hijack our 
+`ResponseWriter`'s `Write` to inspect the call stack and find
+our HTTP handler before performing the `Write`:
+
+```go
+func (rw *ResponseWriter) Write(b []byte) (int, error) {
+	rw.setHandlerInfo()
+	return rw.W.Write(b)
+}
+```
+
+Here, `setHandlerInfo` iterates up the call stack (using Go's `runtime` 
+package) until it finds a caller whose declaration is inside the 
+package we are testing (in all likelihood, the handler).
+
+Once we have the handler's function name, we can get the Go doc string 
+using the `go/doc` package.
+
+
+## Thoughts for the future
+I'd like to convert the main types to interfaces, allowing support for 
+formats other than the API Blueprint format (eg, Google's discovery js).
```

Comment:
To be honest, I have never actually played with Google discovery.js. I recently got a question on twitter asking if I'll support discovery.js in the future, which is why "discover js" is in here. 
I'm going to take out this reference, and leave it at "formats other than the API Blueprint format"